### ISSUE TITLE
[naming-convention] React Functional Components failing variableLike rule

### ISSUE THREAD

**Issue by JoshRobertson** [typescript-eslint/typescript-eslint#2185]
As a part of our migration to ESLint 7, we've updated @typescript-eslint/eslint-plugin from 2.34.0 to 3.1.0, and it's causing React Functional Components to failed variableLike camelCase naming convention rules. This was not the case with the old camelcase rule.

**Repro**

Removed:
```
  extends: [
    "plugin:@typescript-eslint/recommended-requiring-type-checking",
    "plugin:@typescript-eslint/eslint-recommended",
]

rules: {
    "@typescript-eslint/camelcase": "error",
    "@typescript-eslint/class-name-casing": "error",
}
```

Added:
```
    "@typescript-eslint/naming-convention": [
      "error",
      {
        selector: "default",
        format: ["camelCase"],
      },

      {
        selector: "variableLike",
        format: ["camelCase", "UPPER_CASE"],
      },
      {
        selector: "parameter",
        format: ["camelCase"],
        leadingUnderscore: "allow",
      },

      {
        selector: "memberLike",
        modifiers: ["private"],
        format: ["camelCase"],
        leadingUnderscore: "require",
      },

      {
        selector: "typeLike",
        format: ["PascalCase"],
      },
    ],
```

Example Component:

```
import React, { FC } from "react";

const SampleComponent: FC = () => {
  return (
    <div>
      <SvgLogo />
      <Title type="h1" }>
        Some Title
      </Title>
    </div>
  );
};

export { SampleComponent };

```

**Expected Result**

React Functional Components should not flagged as failing variable naming camelCasing rule because React components are required to be PascalCase.

**Actual Result**

React FC are failing the variableLike camelCasing rule

**Additional Info**

```
  error  Variable name SampleComponent must match one of the following formats: camelCase, UPPER_CASE  @typescript-eslint/naming-convention
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.1.0` |
| `@typescript-eslint/parser`        | `3.1.0` |
| `TypeScript`                       | `3.9.5` |
| `ESLint`                           | `7.2.0` |
| `node`                             | `12.16.0` |
| `yarn`                              | `1.22.4` |

**Comment by bradzacher** [comment]
You're declaring a variable, so the name has to match your configuration.

If your codebase requires variables to be named with `PascalCase`, then you should add that to the allowed formats for the `variable` selector.

**Comment by JoshRobertson** [comment]
> If your codebase requires variables to be named with `PascalCase`, then you should add that to the allowed formats for the `variable` selector.

@bradzacher this would be functionally equivalent to disabling linting camelCase in React applications since its a requirement of the framework.

How would you define a customer ruleset for Component names that doesn't require giving up camelCase otherwise?

**Comment by bradzacher** [comment]
Sorry, I was on mobile and was trying to be succinct because of it.

Full response:

> This was not the case with the old camelcase rule.

The configurations provided in the docs to "mimic" the base camelcase ruleset are a little stricter than the base eslint camelcase rule.

This is because the base camelcase rule really didn't do any proper validation, and allowed a name if the following two conditions were met:
- it does not contain an underscore except for in the first or last positions
- if it does contain an underscore, **all** characters must be uppercase

Other than that, it put no other rules on the name - you may have any combination of upper or lowercase.

examples:
```ts
// VALID
const xxxxxxxxxxxxx = 1;
const XXXXXXXXXXXXX = 1;
const XXXXXX_XXXXXX = 1;
const XxXxXxXxXxXxX = 1;
const _xxxxxxxxxxxx = 1;
const xxxxxxxxxxxx_ = 1;
const XxXxXxXxXxXx_ = 1;
const _XxXxXxXxXxXx = 1;

// INVALID
const xxxxxx_xxxxx = 1;
const XxXxXx_XxXxX = 1;
```
[demo on eslint](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IGNhbWVsY2FzZTogWydlcnJvciddICovXG5cbi8vIFZBTElEXG5cbmNvbnN0IHh4eHh4eHh4eHh4eHggPSAxO1xuY29uc3QgWFhYWFhYWFhYWFhYWCA9IDE7XG5jb25zdCBYWFhYWFhfWFhYWFhYID0gMTtcbmNvbnN0IFh4WHhYeFh4WHhYeFggPSAxO1xuY29uc3QgX3h4eHh4eHh4eHh4eCA9IDE7XG5jb25zdCB4eHh4eHh4eHh4eHhfID0gMTtcbmNvbnN0IFh4WHhYeFh4WHhYeF8gPSAxO1xuY29uc3QgX1h4WHhYeFh4WHhYeCA9IDE7XG5cbi8vIElOVkFMSURcblxuY29uc3QgeHh4eHh4X3h4eHh4ID0gMTtcbmNvbnN0IFh4WHhYeF9YeFh4WCA9IDE7XG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjExLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)

This naming convention can be expressed by:
```json5
{
  selector: 'variableLike',
  format: ['camelCase', 'PascalCase', 'UPPER_CASE'],
  leadingUnderscore: 'allow',
  trailingUnderscore: 'allow'
}
```

In the docs, I was a little stricter and omitted `PascalCase`, because it looks wrong and weird to people that `variableLike` would also allow `PascalCase` when it's mimicking the `camelcase` rule.

This is why I mention that you can just add `PascalCase` to your allowed formats.

> How would you define a customer ruleset for Component names that doesn't require giving up camelCase otherwise?

You can limit the scope of `PascalCase` in few ways:

1) define a `selector` specifically on `variable` instead of `variableLike`
  - useful if you're defining functional components as variables.
1) define a `selector` specifically on `variable` with `type` `function`
  - more specific than before, but requires type information, which you may or may not use.
1) define a `selector` specifically on `function` instead of `variableLike`
  - useful if you're defining functional components as functions.
1) use [ESLint `overrides`](https://eslint.org/docs/user-guide/configuring#configuration-based-on-glob-patterns) along with one of the above to further limit the scope of pascal case to `.tsx` files.

For more information on configuring, I suggest perusing the [rule's docs](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/naming-convention.md#selector-options); I tried to be very comprehensive with them, so LMK if anything is missing.
